cmd/util/ledger/migrations/deduplicate_contract_names_migration.go

```diff
@@ -0,0 +1,134 @@
+package migrations
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/onflow/cadence/runtime/common"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/fvm/environment"
+	"github.com/onflow/flow-go/fvm/storage/state"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// DeduplicateContractNamesMigration checks if the contract names have been duplicated and
+// removes the duplicate ones.
+type DeduplicateContractNamesMigration struct {
+	log zerolog.Logger
+}
+
+func (d *DeduplicateContractNamesMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	d.log = log.
+		With().
+		Str("migration", "DeduplicateContractNamesMigration").
+		Logger()
+
+	return nil
+}
+
+func (d *DeduplicateContractNamesMigration) MigrateAccount(
+	ctx context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+	snapshot, err := util.NewPayloadSnapshot(payloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create payload snapshot: %w", err)
+	}
+	transactionState := state.NewTransactionState(snapshot, state.DefaultParameters())
+	accounts := environment.NewAccounts(transactionState)
+	flowAddress := flow.ConvertAddress(address)
+
+	contractNames, err := accounts.GetContractNames(flowAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get contract names: %w", err)
+	}
```

Comment:
Creating `snapshot` can be expensive and time-consuming for large accounts.

Maybe we can avoid creating `snapshot` since we are only interested in one payload with known RegisterID.  

For example, this approach can avoid `snapshot`:
```suggestion
	flowAddress := flow.ConvertAddress(address)
	contractNamesID := flow.ContractNamesRegisterID(flowAddress)

	// Find ContractNames payload
	var contractNamesPayload *ledger.Payload
	var contractNamesPayloadIndex int 
	for i, payload := range payloads {
		key, err := payload.Key()
		if err != nil {
			return nil, err
		}
		id, err := convert.LedgerKeyToRegisterID(key)
		if err != nil {
			return nil, err
		}
		if id == contractNamesID {
			contractNamesPayload = payload
			contractNamesPayloadIndex = i
			break
		}
	}

	if contractNamesPayload == nil {
		return payloads, nil
	}

	var contractNames []string
	err := cbor.Unmarshal(contractNamesPayload.Value(), &contractNames)
	if err != nil {
		return nil, fmt.Errorf("failed to get contract names: %w", err)
	}	
```

Reply:
good point!